lib-injection/sitecustomize.py

```diff
@@ -1,52 +1,74 @@
 """
-This module when included on the PYTHONPATH will install the ddtrace package
-from the locally available wheels that are included in the image.
+This module when included on the PYTHONPATH will update the PYTHONPATH to point to a directory
+containing the ddtrace package compatible with the current Python version and platform.
 """
 import os
 import sys
 
 
-def _configure_ddtrace():
-    # This import has the same effect as ddtrace-run for the current process.
-    import ddtrace.bootstrap.sitecustomize
+debug_mode = os.environ.get("DD_TRACE_DEBUG") in ("true", "1")
 
-    bootstrap_dir = os.path.abspath(os.path.dirname(ddtrace.bootstrap.sitecustomize.__file__))
-    prev_python_path = os.getenv("PYTHONPATH", "")
-    os.environ["PYTHONPATH"] = "%s%s%s" % (bootstrap_dir, os.path.pathsep, prev_python_path)
 
-    # Also insert the bootstrap dir in the path of the current python process.
-    sys.path.insert(0, bootstrap_dir)
-    print("datadog autoinstrumentation: successfully configured python package", file=sys.stderr)
+def _log(msg, *args, level="info"):
+    """Log a message to stderr.
 
+    This function is provided instead of built-in Python logging since we can't rely on any logger
+    being configured.
+    """
+    if not debug_mode and level == "debug":
+        return
+    print("%s:datadog.autoinstrumentation(pid: %d): " % (level.upper(), os.getpid()) + msg % args, file=sys.stderr)
+
+
+script_dir = os.path.dirname(__file__)
+pkgs_path = os.path.join(script_dir, "ddtrace_pkgs")
+_log("ddtrace_pkgs path is %r" % pkgs_path, level="debug")
+_log("ddtrace_pkgs contents: %r" % os.listdir(pkgs_path), level="debug")
+
+try:
+    import ddtrace
+except ModuleNotFoundError:
+    _log("user-installed ddtrace not found, configuring application to use injection site-packages")
+
+    python_version = ".".join(str(i) for i in sys.version_info[:2])
+    site_pkgs_path = os.path.join(pkgs_path, "site-packages-ddtrace-py%s" % python_version)
+    _log("site-packages path is %r" % site_pkgs_path, level="debug")
+    if not os.path.exists(site_pkgs_path):
+        _log("ddtrace site-packages not found in %r" % site_pkgs_path, level="error")
+
+    # Add the custom site-packages directory to the Python path to load the ddtrace package.
+    sys.path.insert(0, site_pkgs_path)
+    _log("sys.path %s" % sys.path, level="debug")
+
+    if site_pkgs_path in sys.path_importer_cache:
+        del sys.path_importer_cache[site_pkgs_path]
```

Comment:
We should add a comment here. This is some subtle bs

Reply:
good call out. We can actually remove this code now since we don't attempt to import the package from the empty site-packages which caused the invalid cache entry. Now we only try to import when the package is present.